arch/risc-v/src/common/riscv_swint.c

```diff
@@ -179,6 +187,31 @@ int riscv_swint(int irq, void *context, void *arg)
   riscv_registerdump(regs);
 #endif
 
+  /* Handle the syscall */
+
+  regs = riscv_handle_syscall(regs);
+
+  /* Report what happened.  That might difficult in the case of a context
+   * switch
+   */
+
+#ifdef CONFIG_DEBUG_SYSCALL_INFO
+  if (regs != CURRENT_REGS)
+    {
+      svcinfo("SWInt Return: Context switch!\n");
+      riscv_registerdump((const uintptr_t *)CURRENT_REGS);
+    }
+  else
+    {
+      svcinfo("SWInt Return: %d\n", regs[REG_A0]);
+    }
+#endif
+
+  return OK;
+}
+
+void *riscv_handle_syscall(uintptr_t *regs)
```

Comment:
how about change riscv_handle_syscall to riscv_do_swint?

Reply:
I think this name is more descriptive, as this does not have anything to do with swint (software interrupt).

Reply:
swint is same as syscall in this context.

Reply:
I'll change the name to riscv_do_swint, or how about riscv_perform_swint ? I don't know what the naming in other platforms is, is do_swint more common ?

Reply:
BTW, why not call riscv_swint from riscv_syscall_dispatch? I don't see any benefit to split riscv_swint into two part and:

1. Call riscv_swint from exception_common
2. But call riscv_handle_syscall from riscv_syscall_dispatch

Reply:
There were two points:

1. int riscv_swint(int irq, void *context, void *arg), irq is not active as the call comes from thread, arg is not used / set
2. The handling of CURRENT_REGS, e.g. DEBUGASSERT(regs && regs == CURRENT_REGS);

Reply:
> There were two points:
> 
> 1. int riscv_swint(int irq, void *context, void *arg), irq is not active as the call comes from thread, arg is not used / set, but then again irq is not used either so it does not matter...

We can pass the fix value here.

> 2. The handling of CURRENT_REGS, e.g. DEBUGASSERT(regs && regs == CURRENT_REGS); The function has 1 global variable so it is not really re-entrant
> 
> riscv_syscall_dispatch is re-entrant, always


it isn't a problem since you need disable interrupt anyway to simulate ECALL.

> 
> If you think that removing the test for regs == CURRENT_REGS is fine, then the function can be combined. To me splitting the logical part (the switch case) out of the already massive function made sense.

I think it's better to call up_doirq just like execption_common once riscv_syscall_dispatch finish to prepare the environment.
From the concept, RISCV kernel mode different from ARM is just because S-mode can't generate the exception to self. So, the  change really needed is simulate ECALL inside ssy_callx, other part should be shared with M-mode and S-mode as much as possible.

Reply:
What about CURRENT_REGS ? That is set by the real interrupt handling logic. So references to that must be removed for S-mode. Otherwise I guess that code can be re-used.

Reply:
Good morning, I took a look at riscv_swint.c and the context switch routines are all dependend on the CURRENT_REGS variable, which is defined as:

`#define CURRENT_REGS (g_current_regs[up_cpu_index()])`

It is a global variable that is dependent on the CPU index too. So no, the function is not truly re-entrant. I guess what CURRENT_REGS is needed for is to pass a reference to the modified context to the interrupt return logic, which is not used when using the trampoline to dispatch a syscall.

As riscv_swint can be entered both from the trampoline as well as interrupt, the value of CURRENT_REGS really needs to be set and cleared.

Reply:
Handling of CURRENT_REGS still needs to be added to the syscall dispatcher, then I think it is ready to call riscv_swint

Reply:
Done

Reply:
Each CPU has a dedicated CURRENT_REGS, so it is  re-entrant from CPU level. For a specific CPU, if RISCV implementation follow ARM here:
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/armv8-m/arm_doirq.c#L58-L62
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/armv8-m/arm_doirq.c#L78-L84
The re-entrant should work again.

Reply:
Yes it can be made to work. I already included handling of CURRENT_REGS. I made it into a C-function trampoline because doing the CURRENT_REGS modification is quite tedious in asm. It adds 1 extra function for S-mode, hope this is OK.